File: pyfenstein3d/engine/field_of_view.py
import math
from .vector2d import Vector2d
from .config import RAY_COUNT
from .config import FOV_ANGLE
from .ray import Ray


class FieldOfView():
    def __init__(self, angle: float):
        self.__ang_abs_min = (FOV_ANGLE * -0.5 + math.pi * 2) % (math.pi * 2)
        self.__ang_abs_max = (FOV_ANGLE * 0.5 + math.pi * 2) % (math.pi * 2)
        self.__vector2d_ang_min = Vector2d(1, 0) ** self.__ang_abs_min
        self.__vector2d_ang_max = Vector2d(1, 0) ** self.__ang_abs_max
        self.__vector2d_ang = Vector2d(1, 0)
        self.__dist = 0
        ray_count = RAY_COUNT - RAY_COUNT % 2
        ray_angle = FOV_ANGLE / (RAY_COUNT - 1)
        self.__rays = [Ray((c * ray_angle) - (FOV_ANGLE / 2)) for c in range(ray_count)]

        self.rot(angle)

    @property
    def ang(self):
        return self.__vector2d_ang.ang

    @property
    def ang_min(self):
        return self.__vector2d_ang_min.ang

    @property
    def ang_max(self):
        return self.__vector2d_ang_max.ang

    @property
    def rays(self):
        return tuple(self.__rays)

    def rot(self, rad: float):
        self.__vector2d_ang = self.__vector2d_ang ** rad
        self.__vector2d_ang_min = self.__vector2d_ang_min ** rad
        self.__vector2d_ang_max = self.__vector2d_ang_max ** rad
        for ray in self.__rays:
            ray.rot(rad)

    def cast(self, player, grid):
        pos = Vector2d(player.x, player.y)
        for ray in self.__rays:
            ray.cast_wall(pos, grid)
            if ray.dist > self.__dist:
                self.__dist = ray.dist

        dist = max([ray.dist for ray in self.__rays])
        doors = grid.get_doors_by_fov(pos, self.ang, FOV_ANGLE, dist)
        for ray in self.__rays:
            ray.cast_doors(pos, doors)
            if ray.dist > self.__dist:
                self.__dist = ray.dist

        dist = max([ray.dist for ray in self.__rays])
        items = grid.get_items_by_fov(pos, self.ang, FOV_ANGLE, dist)
        for ray in self.__rays:
            ray.cast_items(player, items)

File: go/engine/field_of_view.go
package engine

import "math"

// FieldOfView manages a set of rays forming the player's field of view.
type FieldOfView struct {
	vector2dAng    *Vector2d
	vector2dAngMin *Vector2d
	vector2dAngMax *Vector2d
	dist           float64
	rays           []*Ray
}

// NewFieldOfView creates a FieldOfView at the given initial angle,
// using the current RayCount and FovAngle package variables.
func NewFieldOfView(angle float64) *FieldOfView {
	fovAngle := FovAngle
	rayCount := RayCount

	angAbsMin := math.Mod(fovAngle*-0.5+math.Pi*2, math.Pi*2)
	angAbsMax := math.Mod(fovAngle*0.5+math.Pi*2, math.Pi*2)

	fov := &FieldOfView{
		vector2dAng:    NewVector2d(1, 0),
		vector2dAngMin: NewVector2d(1, 0).Rot(angAbsMin),
		vector2dAngMax: NewVector2d(1, 0).Rot(angAbsMax),
	}

	rc := rayCount - rayCount%2
	rayAngle := fovAngle / float64(rayCount-1)
	fov.rays = make([]*Ray, rc)
	for c := 0; c < rc; c++ {
		fov.rays[c] = NewRay(float64(c)*rayAngle - fovAngle/2)
	}

	fov.Rot(angle)
	return fov
}

func (f *FieldOfView) Ang() float64    { return f.vector2dAng.Ang() }
func (f *FieldOfView) AngMin() float64 { return f.vector2dAngMin.Ang() }
func (f *FieldOfView) AngMax() float64 { return f.vector2dAngMax.Ang() }
func (f *FieldOfView) Rays() []*Ray    { return f.rays }

// Rot rotates the entire field of view by rad radians.
func (f *FieldOfView) Rot(rad float64) {
	f.vector2dAng = f.vector2dAng.Rot(rad)
	f.vector2dAngMin = f.vector2dAngMin.Rot(rad)
	f.vector2dAngMax = f.vector2dAngMax.Rot(rad)
	for _, ray := range f.rays {
		ray.Rot(rad)
	}
}

// Cast performs wall, door, and item raycasting from the player's position.
func (f *FieldOfView) Cast(player *Player, grid *ItemGrid) {
	pos := NewVector2d(player.X(), player.Y())

	for _, ray := range f.rays {
		ray.CastWall(pos, grid)
		if ray.Dist() > f.dist {
			f.dist = ray.Dist()
		}
	}

	maxDist := 0.0
	for _, ray := range f.rays {
		if ray.Dist() > maxDist {
			maxDist = ray.Dist()
		}
	}

	doors := grid.GetDoorsByFov(pos, f.Ang(), FovAngle, maxDist)
	for _, ray := range f.rays {
		ray.CastDoors(pos, doors)
		if ray.Dist() > f.dist {
			f.dist = ray.Dist()
		}
	}

	maxDist = 0.0
	for _, ray := range f.rays {
		if ray.Dist() > maxDist {
			maxDist = ray.Dist()
		}
	}

	items := grid.GetItemsByFov(pos, f.Ang(), FovAngle, maxDist)
	for _, ray := range f.rays {
		ray.CastItems(player, items)
	}
}
